xdot/ui/elements.py

```diff
@@ -532,12 +533,80 @@ def is_inside(self, x, y):
         return False
 
     def get_jump(self, x, y):
-        if self.is_inside_begin(x, y):
-            return Jump(self, self.dst.x, self.dst.y, highlight=set([self, self.dst]))
-        if self.is_inside_end(x, y):
-            return Jump(self, self.src.x, self.src.y, highlight=set([self, self.src]))
+        for shape in self.shapes:
+            x1, y1, x2, y2 = shape.bounding
+            if (x1 - self.RADIUS) <= x and x <= (x2 + self.RADIUS) and (y1 - self.RADIUS) <= y and y <= (y2 + self.RADIUS):
+                break
+
+        else:
+            return None
+
+        if self._get_distance(x, y) <= self.RADIUS**2:
+            return Jump(self, self.src.x, self.src.y)
+
         return None
 
+    def _get_distance(self, x, y):
+        result = float('inf')
+
+        for shape in self.shapes:
+            if not isinstance(shape, BezierShape):
```

Comment:
Instead of `isinstance(foo, Foo)` we should add the necessary methods to the classes, so next time we add a new shape, we know how it needs to be extended.